Authentication

jsheng112 opened the pull request:
Added authentication capabilities to website:
- comment form remains hidden until the user login
- toggle between login and logout buttons for the right most nav bar tab
- all blog comments and normal comments now have the "email" property that associates a comment with a specific user

Commit:
add button that returns login status

Commit:
hide the comments for unlogged in users

Commit:
add emails to comments

Commit:
add login/logout buttons

Commit:
add file upload input with Blobstore url

meltsufin commented on portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java:
```diff
@@ -0,0 +1,68 @@
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+
+      // Send the JSON as the response
+      String[] result = {"Stranger", loginUrl};
```
This block of 4 lines is almost identical as the ones in the `if` block above. I would move them out of the `if-else` and re-use them to avoid duplication.

meltsufin commented on portfolio/src/main/webapp/blog.html:
```diff
@@ -7,9 +7,9 @@
     <script src="script.js"></script>
 </head>
-<body onload="setColor();loadPosts();">
+<body onload="setColor();loadPosts();checkAuth('blog')">
```
Since now you're calling 3 functions here. I would replace it with a single function called `init('blog')`, and define it in `script.js`.

meltsufin reviewed (changes requested):
I like how you did navigation bar Login/Logout!
Note that it's still possible to delete comments if you're not logged in.
I'm not sure if that's supposed to be covered in future steps, but I just noticed that when testing your app.

Commit:
fix requested changes for authentication

small changes made to authentication. Delete button now
inaccessible unless user is logged in.

I already have some blogstore code on this branch that is currently
not working. After this round of pull request review I will make sure to
make a new branch.

meltsufin reviewed (dismissed):
The auth part looks good to me.
If you need guidance on how to work with multiple branches at the same time, like keeping the blobstore stuff on a different branch while you continue working on auth, let me know. Otherwise, it's OK that it's in this branch.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java:
```diff
@@ -0,0 +1,64 @@
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
```
Remove this line, since you are setting content type to json later in the method.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java:
```diff
@@ -0,0 +1,64 @@
+    String redirect = request.getParameter("redirect");
+
+    UserService userService = UserServiceFactory.getUserService();
+    String[] result = new String[2];
```
Using a map might be more maintainable, since then you can add new properties without worrying about order.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/BlogCommentServlet.java:
```diff
@@ -55,8 +57,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+      String email = (String) entity.getProperty("email");
 
-      Comment comment = new Comment(content, time, name, postId, commentId, emoji);
+      Comment comment = new Comment(content, time, name, postId, commentId, emoji, email, "");
```
(entirely optional) If you run out of things to do, explore the Builder pattern; here is a good blog post -- https://medium.com/@ajinkyabadve/builder-design-patterns-in-java-1ffb12648850

elefeint commented on portfolio/src/main/webapp/script.js:
```diff
@@ -285,6 +285,10 @@ function createDivElement(comment, isBlogComment) {
   divElement.appendChild(pElementEmoji);
 
+  const imgElement = document.createElement('a');
+  imgElement.innerHTML = "<a href=\"" + comment.image + "\"><img src=\"" + comment.image + "\" /></a>"
```
You can set attributes in the element directly without concatenating HTML.
https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement

elefeint reviewed (changes requested):
Looks great!

Commit:
fix requested changes
